In [13]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [16]:
df = pd.read_csv('/content/drive/MyDrive/Datasets/Decision tree- boston housing/HousingData.csv')

In [17]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [19]:
x=df.drop(['MEDV'], axis=1)
y=df['MEDV']

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     486 non-null    float64
 1   ZN       486 non-null    float64
 2   INDUS    486 non-null    float64
 3   CHAS     486 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      486 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    486 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [21]:
data.describe()

RM       LSTAT     PTRATIO          MEDV
count  489.000000  489.000000  489.000000  4.890000e+02
mean     6.240288   12.939632   18.516564  4.543429e+05
std      0.643650    7.081990    2.111268  1.653403e+05
min      3.561000    1.980000   12.600000  1.050000e+05
25%      5.880000    7.370000   17.400000  3.507000e+05
50%      6.185000   11.690000   19.100000  4.389000e+05
75%      6.575000   17.120000   20.200000  5.187000e+05
max      8.398000   37.970000   22.000000  1.024800e+06

Splitting data

In [22]:
x.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  
0  396.90   4.98  
1  396.90   9.14  
2  392.83   4.03  
3  394.63   2.94  
4  396.90    NaN

Feature scaling

In [24]:
from sklearn.preprocessing import StandardScaler

sc_x= StandardScaler()
x=sc_x.fit_transform(x)

In [25]:
from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest=train_test_split(x,y, test_size=0.3, random_state=42)

In [26]:
xtrain.shape

(354, 13)

In [27]:
xtest.shape

(152, 13)

Decision Tree

In [36]:
from sklearn.impute import SimpleImputer

# Create a SimpleImputer object with the desired strategy
# Strategies include 'mean', 'median', 'most_frequent', or a constant value
imputer = SimpleImputer(strategy='mean')  # You can replace 'mean' with other strategies

# Fit the imputer on the data (compute the mean, median, or most frequent value)
imputer.fit(xtrain)

# Transform the dataset by replacing missing values with the computed values
X_imputed = imputer.transform(xtrain)
X_imputed = imputer.fit_transform(xtrain)



In [52]:
import numpy as np
from sklearn.impute import SimpleImputer
imp_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
xtrain=imp_mean.fit_transform(xtrain)

In [56]:
import numpy as np
from sklearn.impute import SimpleImputer
imp_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
xtest=imp_mean.fit_transform(xtest)

In [53]:
xtrain

array([[-0.41119679, -0.47986431, -1.30387674, ...,  0.1130321 ,
         0.41057102, -1.04993031],
       [-0.39951874, -0.47986431, -0.15727248, ..., -0.30309415,
         0.40158024, -0.03490525],
       [-0.39494533, -0.47986431, -0.61122819, ..., -0.2568579 ,
         0.44105193, -0.35048337],
       ...,
       [-0.41290034,  2.94408951, -1.32877108, ..., -0.67298414,
         0.37570436, -0.94081659],
       [ 0.86652956, -0.47986431,  1.02740548, ...,  0.80657583,
        -2.70626713,  1.47647206],
       [-0.38836301, -0.47986431, -0.3696066 , ...,  1.13022958,
        -3.13442533, -0.29172982]])

In [54]:
from sklearn.tree import DecisionTreeRegressor

dtr=DecisionTreeRegressor()
dtr.fit(xtrain, ytrain)

DecisionTreeRegressor()

In [57]:
y_pred=dtr.predict(xtest)

In [58]:
y_pred[:10]

array([21.6, 33.1,  8.1, 24.3, 17.7, 21.4, 16.6, 13.3, 20.2, 11.9])

In [59]:
ytest[:10]

173    23.6
274    32.4
491    13.6
72     22.8
452    16.1
76     20.0
316    17.8
140    14.0
471    19.6
500    16.8
Name: MEDV, dtype: float64

In [60]:
from sklearn.metrics import mean_squared_error

In [61]:
mean_squared_error(ytest, y_pred)

16.735526315789475

16 % error in prediction

In [62]:
from sklearn.ensemble import RandomForestRegressor

rf_reg=RandomForestRegressor(n_estimators=500)
rf_reg.fit(xtrain, ytrain)

RandomForestRegressor(n_estimators=500)

In [63]:
y_pred=rf_reg.predict(xtest)

In [64]:
mean_squared_error(ytest, y_pred)

9.198389167105248

9.1% error

Random forest muchh better than decision tree